File: sdk/js/src/sporepay.ts
import { Connection, ReadonlyContractResult } from './connection.js';
import { Keypair } from './keypair.js';
import { PublicKey } from './publickey.js';

const PROGRAM_SYMBOL_CANDIDATES = ['SPOREPAY', 'sporepay'];
const STREAM_SIZE = 105;
const STREAM_INFO_SIZE = 113;
const MAX_U64 = (1n << 64n) - 1n;

export interface SporePayStream {
    streamId: bigint;
    sender: string;
    recipient: string;
    totalAmount: bigint;
    withdrawnAmount: bigint;
    startSlot: bigint;
    endSlot: bigint;
    cancelled: boolean;
    createdSlot: bigint;
}

export interface SporePayStreamInfo extends SporePayStream {
    cliffSlot: bigint;
}

export interface SporePayStreamIdPage {
    totalCount: bigint;
    nextCursor: bigint;
    streamIds: bigint[];
}

export interface SporePayStats {
    streamCount: number;
    totalStreamed: number;
    totalWithdrawn: number;
    cancelCount: number;
    escrowLiability: number;
    unpaidLiability: number;
    accountingVersion: number;
    migrationLocked: boolean;
    migrationExpectedStreams: number;
    migrationCursor: number;
    paused: boolean;
}

export interface CreateStreamParams {
    recipient: PublicKey | string;
    totalAmount: number | bigint;
    startSlot: number | bigint;
    endSlot: number | bigint;
}

export interface CreateStreamWithCliffParams extends CreateStreamParams {
    cliffSlot: number | bigint;
}

export interface WithdrawFromStreamParams {
    streamId: number | bigint;
    amount: number | bigint;
}

export interface TransferStreamParams {
    streamId: number | bigint;
    newRecipient: PublicKey | string;
}

function normalizeAddress(value: PublicKey | string): PublicKey {
    return value instanceof PublicKey ? value : new PublicKey(value);
}

function normalizeUnsignedU64(value: number | bigint, fieldName: string): bigint {
    const normalized = typeof value === 'bigint'
        ? value
        : Number.isSafeInteger(value) && value >= 0
            ? BigInt(value)
            : null;

    if (normalized === null || normalized < 0n || normalized > MAX_U64) {
        throw new Error(`${fieldName} must be a u64-safe integer value`);
    }

    return normalized;
}

function u64LE(value: number | bigint, fieldName: string): Uint8Array {
    const out = new Uint8Array(8);
    new DataView(out.buffer).setBigUint64(0, normalizeUnsignedU64(value, fieldName), true);
    return out;
}

function buildLayoutArgs(layout: number[], chunks: Uint8Array[]): Uint8Array {
    const header = Uint8Array.from([0xAB, ...layout]);
    const total = chunks.reduce((sum, chunk) => sum + chunk.length, header.length);
    const out = new Uint8Array(total);
    out.set(header, 0);
    let offset = header.length;
    for (const chunk of chunks) {
        out.set(chunk, offset);
        offset += chunk.length;
    }
    return out;
}

function encodeCreateStreamArgs(sender: PublicKey, params: CreateStreamParams): Uint8Array {
    return buildLayoutArgs([0x20, 0x20, 0x08, 0x08, 0x08], [
        sender.toBytes(),
        normalizeAddress(params.recipient).toBytes(),
        u64LE(params.totalAmount, 'totalAmount'),
        u64LE(params.startSlot, 'startSlot'),
        u64LE(params.endSlot, 'endSlot'),
    ]);
}

function encodeCreateStreamWithCliffArgs(sender: PublicKey, params: CreateStreamWithCliffParams): Uint8Array {
    return buildLayoutArgs([0x20, 0x20, 0x08, 0x08, 0x08, 0x08], [
        sender.toBytes(),
        normalizeAddress(params.recipient).toBytes(),
        u64LE(params.totalAmount, 'totalAmount'),
        u64LE(params.startSlot, 'startSlot'),
        u64LE(params.endSlot, 'endSlot'),
        u64LE(params.cliffSlot, 'cliffSlot'),
    ]);
}

function encodeWithdrawArgs(caller: PublicKey, params: WithdrawFromStreamParams): Uint8Array {
    return buildLayoutArgs([0x20, 0x08, 0x08], [
        caller.toBytes(),
        u64LE(params.streamId, 'streamId'),
        u64LE(params.amount, 'amount'),
    ]);
}

function encodeCancelArgs(caller: PublicKey, streamId: number | bigint): Uint8Array {
    return buildLayoutArgs([0x20, 0x08], [
        caller.toBytes(),
        u64LE(streamId, 'streamId'),
    ]);
}

function encodeTransferArgs(caller: PublicKey, params: TransferStreamParams): Uint8Array {
    return buildLayoutArgs([0x20, 0x20, 0x08], [
        caller.toBytes(),
        normalizeAddress(params.newRecipient).toBytes(),
        u64LE(params.streamId, 'streamId'),
    ]);
}

function encodeStreamLookupArgs(streamId: number | bigint): Uint8Array {
    return buildLayoutArgs([0x08], [u64LE(streamId, 'streamId')]);
}

function encodeAddressArgs(address: PublicKey): Uint8Array {
    return buildLayoutArgs([0x20], [address.toBytes()]);
}

function encodeAddressPageArgs(
    address: PublicKey,
    cursor: number | bigint,
    limit: number | bigint,
): Uint8Array {
    return buildLayoutArgs([0x20, 0x08, 0x08], [
        address.toBytes(),
        u64LE(cursor, 'cursor'),
        u64LE(limit, 'limit'),
    ]);
}

function decodeReturnData(returnData: string): Uint8Array {
    return Uint8Array.from(Buffer.from(returnData, 'base64'));
}

function readU64(bytes: Uint8Array, offset: number): bigint {
    const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    return view.getBigUint64(offset, true);
}

function ensureReturnCodeZero(result: ReadonlyContractResult, functionName: string): void {
    const code = result.returnCode ?? 0;
    if (code !== 0) {
        throw new Error(result.error ?? `SporePay ${functionName} returned code ${code}`);
    }
    if (result.success === false && result.error) {
        throw new Error(result.error);
    }
}

function decodeStream(streamId: bigint, bytes: Uint8Array): SporePayStream {
    if (bytes.length < STREAM_SIZE) {
        throw new Error('SporePay stream payload was shorter than expected');
    }

    return {
        streamId,
        sender: PublicKey.fromBytes(bytes.slice(0, 32)).toBase58(),
        recipient: PublicKey.fromBytes(bytes.slice(32, 64)).toBase58(),
        totalAmount: readU64(bytes, 64),
        withdrawnAmount: readU64(bytes, 72),
        startSlot: readU64(bytes, 80),
        endSlot: readU64(bytes, 88),
        cancelled: bytes[96] === 1,
        createdSlot: readU64(bytes, 97),
    };
}

function decodeStreamInfo(streamId: bigint, bytes: Uint8Array): SporePayStreamInfo {
    if (bytes.length < STREAM_INFO_SIZE) {
        throw new Error('SporePay stream-info payload was shorter than expected');
    }

    return {
        ...decodeStream(streamId, bytes),
        cliffSlot: readU64(bytes, 105),
    };
}

function decodeStreamIdPage(bytes: Uint8Array): SporePayStreamIdPage {
    if (bytes.length < 24) {
        throw new Error('SporePay stream-ID page payload was shorter than expected');
    }
    const returned = readU64(bytes, 16);
    if (returned > BigInt(Number.MAX_SAFE_INTEGER)
        || bytes.length !== 24 + Number(returned) * 8) {
        throw new Error('SporePay stream-ID page payload length was inconsistent');
    }
    const streamIds: bigint[] = [];
    for (let index = 0; index < Number(returned); index += 1) {
        streamIds.push(readU64(bytes, 24 + index * 8));
    }
    return {
        totalCount: readU64(bytes, 0),
        nextCursor: readU64(bytes, 8),
        streamIds,
    };
}

export class SporePayClient {
    private resolvedProgram?: PublicKey;

    constructor(
        private readonly connection: Connection,
        programId?: PublicKey,
    ) {
        this.resolvedProgram = programId;
    }

    private async callReadonly(functionName: string, args: Uint8Array): Promise<ReadonlyContractResult> {
        const programId = await this.getProgramId();
        return this.connection.callReadonlyContract(programId, functionName, args);
    }

    async getProgramId(): Promise<PublicKey> {
        if (this.resolvedProgram) {
            return this.resolvedProgram;
        }

        for (const symbol of PROGRAM_SYMBOL_CANDIDATES) {
            try {
                const entry = await this.connection.getSymbolRegistry(symbol);
                if (entry?.program) {
                    this.resolvedProgram = new PublicKey(entry.program);
                    return this.resolvedProgram;
                }
            } catch {
                // Try the next known registry alias.
            }
        }

        throw new Error('Unable to resolve the SporePay program via getSymbolRegistry("SPOREPAY")');
    }

    async getStream(streamId: number | bigint): Promise<SporePayStream | null> {
        const normalizedStreamId = normalizeUnsignedU64(streamId, 'streamId');
        const result = await this.callReadonly('get_stream', encodeStreamLookupArgs(normalizedStreamId));
        if ((result.returnCode ?? 0) === 1 || !result.returnData) {
            return null;
        }
        ensureReturnCodeZero(result, 'get_stream');
        return decodeStream(normalizedStreamId, decodeReturnData(result.returnData));
    }

    async getStreamInfo(streamId: number | bigint): Promise<SporePayStreamInfo | null> {
        const normalizedStreamId = normalizeUnsignedU64(streamId, 'streamId');
        const result = await this.callReadonly('get_stream_info', encodeStreamLookupArgs(normalizedStreamId));
        if ((result.returnCode ?? 0) === 1 || !result.returnData) {
            return null;
        }
        ensureReturnCodeZero(result, 'get_stream_info');
        return decodeStreamInfo(normalizedStreamId, decodeReturnData(result.returnData));
    }

    async getWithdrawable(streamId: number | bigint): Promise<bigint> {
        const result = await this.callReadonly('get_withdrawable', encodeStreamLookupArgs(streamId));
        ensureReturnCodeZero(result, 'get_withdrawable');
        if (!result.returnData) {
            throw new Error('SporePay get_withdrawable did not return a balance');
        }
        return readU64(decodeReturnData(result.returnData), 0);
    }

    async getStats(): Promise<SporePayStats> {
        const stats = await this.connection.getSporePayStats();
        return {
            streamCount: stats.stream_count ?? 0,
            totalStreamed: stats.total_streamed ?? 0,
            totalWithdrawn: stats.total_withdrawn ?? 0,
            cancelCount: stats.cancel_count ?? 0,
            escrowLiability: stats.escrow_liability ?? 0,
            unpaidLiability: stats.unpaid_liability ?? 0,
            accountingVersion: stats.accounting_version ?? 0,
            migrationLocked: Boolean(stats.migration_locked),
            migrationExpectedStreams: stats.migration_expected_streams ?? 0,
            migrationCursor: stats.migration_cursor ?? 0,
            paused: Boolean(stats.paused),
        };
    }

    async getUnpaidPayout(recipient: PublicKey | string): Promise<bigint> {
        const result = await this.callReadonly(
            'get_unpaid_payout',
            encodeAddressArgs(normalizeAddress(recipient)),
        );
        ensureReturnCodeZero(result, 'get_unpaid_payout');
        if (!result.returnData) {
            throw new Error('SporePay get_unpaid_payout did not return a balance');
        }
        return readU64(decodeReturnData(result.returnData), 0);
    }

    async getSenderStreamIds(
        sender: PublicKey | string,
        cursor: number | bigint = 0,
        limit: number | bigint = 64,
    ): Promise<SporePayStreamIdPage> {
        const result = await this.callReadonly(
            'get_sender_stream_ids',
            encodeAddressPageArgs(normalizeAddress(sender), cursor, limit),
        );
        ensureReturnCodeZero(result, 'get_sender_stream_ids');
        if (!result.returnData) {
            throw new Error('SporePay get_sender_stream_ids did not return a page');
        }
        return decodeStreamIdPage(decodeReturnData(result.returnData));
    }

    async getRecipientStreamIds(
        recipient: PublicKey | string,
        cursor: number | bigint = 0,
        limit: number | bigint = 64,
    ): Promise<SporePayStreamIdPage> {
        const result = await this.callReadonly(
            'get_recipient_stream_ids',
            encodeAddressPageArgs(normalizeAddress(recipient), cursor, limit),
        );
        ensureReturnCodeZero(result, 'get_recipient_stream_ids');
        if (!result.returnData) {
            throw new Error('SporePay get_recipient_stream_ids did not return a page');
        }
        return decodeStreamIdPage(decodeReturnData(result.returnData));
    }

    async createStream(sender: Keypair, params: CreateStreamParams): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeCreateStreamArgs(sender.pubkey(), params);
        return this.connection.callContract(sender, programId, 'create_stream', args);
    }

    async createStreamWithCliff(sender: Keypair, params: CreateStreamWithCliffParams): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeCreateStreamWithCliffArgs(sender.pubkey(), params);
        return this.connection.callContract(sender, programId, 'create_stream_with_cliff', args);
    }

    async withdrawFromStream(recipient: Keypair, params: WithdrawFromStreamParams): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeWithdrawArgs(recipient.pubkey(), params);
        return this.connection.callContract(recipient, programId, 'withdraw_from_stream', args);
    }

    async cancelStream(sender: Keypair, streamId: number | bigint): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeCancelArgs(sender.pubkey(), streamId);
        return this.connection.callContract(sender, programId, 'cancel_stream', args);
    }

    async transferStream(recipient: Keypair, params: TransferStreamParams): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeTransferArgs(recipient.pubkey(), params);
        return this.connection.callContract(recipient, programId, 'transfer_stream', args);
    }

    async claimUnpaidPayout(recipient: Keypair): Promise<string> {
        const programId = await this.getProgramId();
        return this.connection.callContract(
            recipient,
            programId,
            'claim_unpaid_payout',
            encodeAddressArgs(recipient.pubkey()),
        );
    }
}

File: sdk/rust/src/sporepay.rs
use crate::{Client, Error, Keypair, Pubkey, Result};
use serde::{Deserialize, Serialize};

const PROGRAM_SYMBOL_CANDIDATES: [&str; 2] = ["SPOREPAY", "sporepay"];
const STREAM_SIZE: usize = 105;
const STREAM_INFO_SIZE: usize = 113;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SporePayStream {
    pub stream_id: u64,
    pub sender: String,
    pub recipient: String,
    pub total_amount: u64,
    pub withdrawn_amount: u64,
    pub start_slot: u64,
    pub end_slot: u64,
    pub cancelled: bool,
    pub created_slot: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SporePayStreamInfo {
    #[serde(flatten)]
    pub stream: SporePayStream,
    pub cliff_slot: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SporePayStreamIdPage {
    pub total_count: u64,
    pub next_cursor: u64,
    pub stream_ids: Vec<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SporePayStats {
    pub stream_count: u64,
    pub total_streamed: u64,
    pub total_withdrawn: u64,
    pub cancel_count: u64,
    pub escrow_liability: u64,
    pub unpaid_liability: u64,
    pub accounting_version: u64,
    pub migration_locked: bool,
    pub migration_expected_streams: u64,
    pub migration_cursor: u64,
    pub paused: bool,
}

#[derive(Debug, Clone)]
pub struct CreateStreamParams {
    pub recipient: Pubkey,
    pub total_amount: u64,
    pub start_slot: u64,
    pub end_slot: u64,
}

#[derive(Debug, Clone)]
pub struct CreateStreamWithCliffParams {
    pub recipient: Pubkey,
    pub total_amount: u64,
    pub start_slot: u64,
    pub end_slot: u64,
    pub cliff_slot: u64,
}

#[derive(Debug, Clone)]
pub struct WithdrawFromStreamParams {
    pub stream_id: u64,
    pub amount: u64,
}

#[derive(Debug, Clone)]
pub struct TransferStreamParams {
    pub stream_id: u64,
    pub new_recipient: Pubkey,
}

#[derive(Debug, Clone)]
pub struct SporePayClient {
    client: Client,
    program_id: std::sync::Arc<std::sync::Mutex<Option<Pubkey>>>,
}

fn build_layout_args(layout: &[u8], chunks: &[Vec<u8>]) -> Vec<u8> {
    let mut out = Vec::with_capacity(
        1 + layout.len() + chunks.iter().map(|chunk| chunk.len()).sum::<usize>(),
    );
    out.push(0xAB);
    out.extend_from_slice(layout);
    for chunk in chunks {
        out.extend_from_slice(chunk);
    }
    out
}

fn encode_create_stream_args(sender: &Pubkey, params: &CreateStreamParams) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x20, 0x08, 0x08, 0x08],
        &[
            sender.as_ref().to_vec(),
            params.recipient.as_ref().to_vec(),
            params.total_amount.to_le_bytes().to_vec(),
            params.start_slot.to_le_bytes().to_vec(),
            params.end_slot.to_le_bytes().to_vec(),
        ],
    )
}

fn encode_create_stream_with_cliff_args(
    sender: &Pubkey,
    params: &CreateStreamWithCliffParams,
) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x20, 0x08, 0x08, 0x08, 0x08],
        &[
            sender.as_ref().to_vec(),
            params.recipient.as_ref().to_vec(),
            params.total_amount.to_le_bytes().to_vec(),
            params.start_slot.to_le_bytes().to_vec(),
            params.end_slot.to_le_bytes().to_vec(),
            params.cliff_slot.to_le_bytes().to_vec(),
        ],
    )
}

fn encode_withdraw_args(caller: &Pubkey, params: &WithdrawFromStreamParams) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x08, 0x08],
        &[
            caller.as_ref().to_vec(),
            params.stream_id.to_le_bytes().to_vec(),
            params.amount.to_le_bytes().to_vec(),
        ],
    )
}

fn encode_cancel_args(caller: &Pubkey, stream_id: u64) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x08],
        &[caller.as_ref().to_vec(), stream_id.to_le_bytes().to_vec()],
    )
}

fn encode_transfer_args(caller: &Pubkey, params: &TransferStreamParams) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x20, 0x08],
        &[
            caller.as_ref().to_vec(),
            params.new_recipient.as_ref().to_vec(),
            params.stream_id.to_le_bytes().to_vec(),
        ],
    )
}

fn encode_stream_lookup_args(stream_id: u64) -> Vec<u8> {
    build_layout_args(&[0x08], &[stream_id.to_le_bytes().to_vec()])
}

fn encode_address_args(address: &Pubkey) -> Vec<u8> {
    build_layout_args(&[0x20], &[address.as_ref().to_vec()])
}

fn encode_address_page_args(address: &Pubkey, cursor: u64, limit: u64) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x08, 0x08],
        &[
            address.as_ref().to_vec(),
            cursor.to_le_bytes().to_vec(),
            limit.to_le_bytes().to_vec(),
        ],
    )
}

fn ensure_readonly_success(
    result: &crate::client::ReadonlyContractResult,
    function_name: &str,
) -> Result<()> {
    let code = result.return_code.unwrap_or(0);
    if code != 0 {
        return Err(Error::RpcError(result.error.clone().unwrap_or_else(|| {
            format!("SporePay {} returned code {}", function_name, code)
        })));
    }
    if !result.success {
        return Err(Error::RpcError(
            result
                .error
                .clone()
                .unwrap_or_else(|| format!("SporePay {} failed", function_name)),
        ));
    }
    Ok(())
}

fn decode_return_data(
    result: &crate::client::ReadonlyContractResult,
    function_name: &str,
) -> Result<Vec<u8>> {
    let Some(return_data) = &result.return_data else {
        return Err(Error::ParseError(format!(
            "SporePay {} did not return payload data",
            function_name,
        )));
    };

    base64::Engine::decode(&base64::engine::general_purpose::STANDARD, return_data)
        .map_err(|err| Error::ParseError(err.to_string()))
}

fn decode_stream(stream_id: u64, bytes: &[u8]) -> Result<SporePayStream> {
    if bytes.len() < STREAM_SIZE {
        return Err(Error::ParseError(
            "SporePay stream payload was shorter than expected".into(),
        ));
    }

    Ok(SporePayStream {
        stream_id,
        sender: Pubkey(bytes[0..32].try_into().unwrap()).to_base58(),
        recipient: Pubkey(bytes[32..64].try_into().unwrap()).to_base58(),
        total_amount: u64::from_le_bytes(bytes[64..72].try_into().unwrap()),
        withdrawn_amount: u64::from_le_bytes(bytes[72..80].try_into().unwrap()),
        start_slot: u64::from_le_bytes(bytes[80..88].try_into().unwrap()),
        end_slot: u64::from_le_bytes(bytes[88..96].try_into().unwrap()),
        cancelled: bytes[96] == 1,
        created_slot: u64::from_le_bytes(bytes[97..105].try_into().unwrap()),
    })
}

fn decode_stream_info(stream_id: u64, bytes: &[u8]) -> Result<SporePayStreamInfo> {
    if bytes.len() < STREAM_INFO_SIZE {
        return Err(Error::ParseError(
            "SporePay stream-info payload was shorter than expected".into(),
        ));
    }

    Ok(SporePayStreamInfo {
        stream: decode_stream(stream_id, bytes)?,
        cliff_slot: u64::from_le_bytes(bytes[105..113].try_into().unwrap()),
    })
}

fn decode_stream_id_page(bytes: &[u8]) -> Result<SporePayStreamIdPage> {
    if bytes.len() < 24 {
        return Err(Error::ParseError(
            "SporePay stream-ID page payload was shorter than expected".into(),
        ));
    }
    let returned = u64::from_le_bytes(bytes[16..24].try_into().unwrap());
    let returned = usize::try_from(returned)
        .map_err(|_| Error::ParseError("SporePay stream-ID page count overflow".into()))?;
    let expected_len =
        24usize
            .checked_add(returned.checked_mul(8).ok_or_else(|| {
                Error::ParseError("SporePay stream-ID page length overflow".into())
            })?)
            .ok_or_else(|| Error::ParseError("SporePay stream-ID page length overflow".into()))?;
    if bytes.len() != expected_len {
        return Err(Error::ParseError(
            "SporePay stream-ID page payload length was inconsistent".into(),
        ));
    }
    let mut stream_ids = Vec::with_capacity(returned);
    for index in 0..returned {
        let offset = 24 + index * 8;
        stream_ids.push(u64::from_le_bytes(
            bytes[offset..offset + 8].try_into().unwrap(),
        ));
    }
    Ok(SporePayStreamIdPage {
        total_count: u64::from_le_bytes(bytes[0..8].try_into().unwrap()),
        next_cursor: u64::from_le_bytes(bytes[8..16].try_into().unwrap()),
        stream_ids,
    })
}

impl SporePayClient {
    pub fn new(client: Client) -> Self {
        Self {
            client,
            program_id: std::sync::Arc::new(std::sync::Mutex::new(None)),
        }
    }

    pub fn with_program_id(client: Client, program_id: Pubkey) -> Self {
        Self {
            client,
            program_id: std::sync::Arc::new(std::sync::Mutex::new(Some(program_id))),
        }
    }

    pub async fn get_program_id(&self) -> Result<Pubkey> {
        if let Some(program_id) = *self
            .program_id
            .lock()
            .map_err(|_| Error::ConfigError("SporePayClient program cache lock poisoned".into()))?
        {
            return Ok(program_id);
        }

        for symbol in PROGRAM_SYMBOL_CANDIDATES {
            let entry = match self.client.get_symbol_registry(symbol).await {
                Ok(entry) => entry,
                Err(_) => continue,
            };
            let Some(program) = entry.get("program").and_then(|value| value.as_str()) else {
                continue;
            };
            let program_id = Pubkey::from_base58(program).map_err(Error::ParseError)?;
            *self.program_id.lock().map_err(|_| {
                Error::ConfigError("SporePayClient program cache lock poisoned".into())
            })? = Some(program_id);
            return Ok(program_id);
        }

        Err(Error::ConfigError(
            "Unable to resolve the SporePay program via getSymbolRegistry(\"SPOREPAY\")".into(),
        ))
    }

    pub async fn get_stream(&self, stream_id: u64) -> Result<Option<SporePayStream>> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_stream",
                encode_stream_lookup_args(stream_id),
                None,
            )
            .await?;

        if result.return_code == Some(1) || result.return_data.is_none() {
            return Ok(None);
        }

        ensure_readonly_success(&result, "get_stream")?;
        let bytes = decode_return_data(&result, "get_stream")?;
        decode_stream(stream_id, &bytes).map(Some)
    }

    pub async fn get_stream_info(&self, stream_id: u64) -> Result<Option<SporePayStreamInfo>> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_stream_info",
                encode_stream_lookup_args(stream_id),
                None,
            )
            .await?;

        if result.return_code == Some(1) || result.return_data.is_none() {
            return Ok(None);
        }

        ensure_readonly_success(&result, "get_stream_info")?;
        let bytes = decode_return_data(&result, "get_stream_info")?;
        decode_stream_info(stream_id, &bytes).map(Some)
    }

    pub async fn get_withdrawable(&self, stream_id: u64) -> Result<u64> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_withdrawable",
                encode_stream_lookup_args(stream_id),
                None,
            )
            .await?;

        ensure_readonly_success(&result, "get_withdrawable")?;
        let bytes = decode_return_data(&result, "get_withdrawable")?;
        if bytes.len() < 8 {
            return Err(Error::ParseError(
                "SporePay withdrawable payload was shorter than expected".into(),
            ));
        }
        Ok(u64::from_le_bytes(bytes[..8].try_into().unwrap()))
    }

    pub async fn get_stats(&self) -> Result<SporePayStats> {
        let value = self.client.get_sporepay_stats().await?;
        serde_json::from_value(value).map_err(|err| Error::ParseError(err.to_string()))
    }

    pub async fn get_unpaid_payout(&self, recipient: &Pubkey) -> Result<u64> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_unpaid_payout",
                encode_address_args(recipient),
                None,
            )
            .await?;
        ensure_readonly_success(&result, "get_unpaid_payout")?;
        let bytes = decode_return_data(&result, "get_unpaid_payout")?;
        if bytes.len() < 8 {
            return Err(Error::ParseError(
                "SporePay unpaid-payout payload was shorter than expected".into(),
            ));
        }
        Ok(u64::from_le_bytes(bytes[..8].try_into().unwrap()))
    }

    async fn get_stream_id_page(
        &self,
        function_name: &str,
        address: &Pubkey,
        cursor: u64,
        limit: u64,
    ) -> Result<SporePayStreamIdPage> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                function_name,
                encode_address_page_args(address, cursor, limit),
                None,
            )
            .await?;
        ensure_readonly_success(&result, function_name)?;
        decode_stream_id_page(&decode_return_data(&result, function_name)?)
    }

    pub async fn get_sender_stream_ids(
        &self,
        sender: &Pubkey,
        cursor: u64,
        limit: u64,
    ) -> Result<SporePayStreamIdPage> {
        self.get_stream_id_page("get_sender_stream_ids", sender, cursor, limit)
            .await
    }

    pub async fn get_recipient_stream_ids(
        &self,
        recipient: &Pubkey,
        cursor: u64,
        limit: u64,
    ) -> Result<SporePayStreamIdPage> {
        self.get_stream_id_page("get_recipient_stream_ids", recipient, cursor, limit)
            .await
    }

    pub async fn create_stream(
        &self,
        sender: &Keypair,
        params: CreateStreamParams,
    ) -> Result<String> {
        let program_id = self.get_program_id().await?;
        let args = encode_create_stream_args(&sender.pubkey(), &params);
        self.client
            .call_contract(sender, &program_id, "create_stream", args, 0)
            .await
    }

    pub async fn create_stream_with_cliff(
        &self,
        sender: &Keypair,
        params: CreateStreamWithCliffParams,
    ) -> Result<String> {
        let program_id = self.get_program_id().await?;
        let args = encode_create_stream_with_cliff_args(&sender.pubkey(), &params);
        self.client
            .call_contract(sender, &program_id, "create_stream_with_cliff", args, 0)
            .await
    }

    pub async fn withdraw_from_stream(
        &self,
        recipient: &Keypair,
        params: WithdrawFromStreamParams,
    ) -> Result<String> {
        let program_id = self.get_program_id().await?;
        let args = encode_withdraw_args(&recipient.pubkey(), &params);
        self.client
            .call_contract(recipient, &program_id, "withdraw_from_stream", args, 0)
            .await
    }

    pub async fn cancel_stream(&self, sender: &Keypair, stream_id: u64) -> Result<String> {
        let program_id = self.get_program_id().await?;
        let args = encode_cancel_args(&sender.pubkey(), stream_id);
        self.client
            .call_contract(sender, &program_id, "cancel_stream", args, 0)
            .await
    }

    pub async fn transfer_stream(
        &self,
        recipient: &Keypair,
        params: TransferStreamParams,
    ) -> Result<String> {
        let program_id = self.get_program_id().await?;
        let args = encode_transfer_args(&recipient.pubkey(), &params);
        self.client
            .call_contract(recipient, &program_id, "transfer_stream", args, 0)
            .await
    }

    pub async fn claim_unpaid_payout(&self, recipient: &Keypair) -> Result<String> {
        let program_id = self.get_program_id().await?;
        self.client
            .call_contract(
                recipient,
                &program_id,
                "claim_unpaid_payout",
                encode_address_args(&recipient.pubkey()),
                0,
            )
            .await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn create_stream_encoding_includes_all_u64_layout_entries() {
        let sender = Pubkey([1u8; 32]);
        let encoded = encode_create_stream_args(
            &sender,
            &CreateStreamParams {
                recipient: Pubkey([2u8; 32]),
                total_amount: 100,
                start_slot: 10,
                end_slot: 20,
            },
        );
        assert_eq!(&encoded[..6], &[0xAB, 0x20, 0x20, 0x08, 0x08, 0x08]);
    }

    #[test]
    fn decode_stream_info_reads_cliff_slot() {
        let mut payload = vec![0u8; STREAM_INFO_SIZE];
        payload[64..72].copy_from_slice(&100u64.to_le_bytes());
        payload[72..80].copy_from_slice(&25u64.to_le_bytes());
        payload[80..88].copy_from_slice(&10u64.to_le_bytes());
        payload[88..96].copy_from_slice(&20u64.to_le_bytes());
        payload[97..105].copy_from_slice(&9u64.to_le_bytes());
        payload[105..113].copy_from_slice(&12u64.to_le_bytes());

        let info = decode_stream_info(7, &payload).unwrap();
        assert_eq!(info.stream.stream_id, 7);
        assert_eq!(info.cliff_slot, 12);
    }

    #[test]
    fn decode_stream_id_page_validates_count_and_length() {
        let mut payload = Vec::new();
        payload.extend_from_slice(&2u64.to_le_bytes());
        payload.extend_from_slice(&2u64.to_le_bytes());
        payload.extend_from_slice(&2u64.to_le_bytes());
        payload.extend_from_slice(&4u64.to_le_bytes());
        payload.extend_from_slice(&9u64.to_le_bytes());
        let page = decode_stream_id_page(&payload).unwrap();
        assert_eq!(page.total_count, 2);
        assert_eq!(page.next_cursor, 2);
        assert_eq!(page.stream_ids, vec![4, 9]);
        assert!(decode_stream_id_page(&payload[..payload.len() - 1]).is_err());
    }
}
